src/app/ngrx/reducers/episode-metrics.reducer.ts

```diff
@@ -21,28 +22,57 @@ const episodeIndex = (state: EpisodeMetricsModel[], id: number, seriesId: number
   return state.findIndex(e => e.seriesId === seriesId && e.id === id);
 };
 
-export function EpisodeMetricsReducer(state: EpisodeMetricsModel[] = initialState, action: AllActions) {
+export function EpisodeMetricsReducer(state: EpisodeMetricsModel[] = initialState, action: ACTIONS.AllActions) {
   switch (action.type) {
-    case ActionTypes.CASTLE_EPISODE_METRICS:
-      if (action instanceof CastleEpisodeMetricsAction) {
-        const {id, seriesId, guid, page} = action.payload.episode;
-        const { metricsPropertyName } = action.payload;
+    case ACTIONS.ActionTypes.CASTLE_EPISODE_METRICS_LOAD:
+      if (action instanceof ACTIONS.CastleEpisodeMetricsLoadAction) {
+        const { id, seriesId, guid, page } = action.payload;
         const epIdx = episodeIndex(state, id, seriesId);
         let episode: EpisodeMetricsModel, newState: EpisodeMetricsModel[];
         if (epIdx > -1) {
-          episode = {...state[epIdx], id, seriesId, guid, page};
-          episode[metricsPropertyName] = action.payload.metrics;
+          episode = {...state[epIdx], id, seriesId, guid, page, loading: true, loaded: false};
           newState = [...state.slice(0, epIdx), episode, ...state.slice(epIdx + 1)];
         } else {
-          episode = {seriesId, id, guid, page};
-          episode[metricsPropertyName] = action.payload.metrics;
+          episode = {seriesId, id, guid, page, loading: true, loaded: false};
           newState = [episode, ...state];
         }
         return newState;
       }
       break;
-    case ActionTypes.CASTLE_EPISODE_CHART_TOGGLE:
-      if (action instanceof CastleEpisodeChartToggleAction) {
+    case ACTIONS.ActionTypes.CASTLE_EPISODE_METRICS_SUCCESS:
+      if (action instanceof ACTIONS.CastleEpisodeMetricsSuccessAction) {
+        const { id, seriesId, guid, page, metricsPropertyName, metrics } = action.payload;
+        const epIdx = episodeIndex(state, id, seriesId);
+        let episode: EpisodeMetricsModel, newState: EpisodeMetricsModel[];
+        if (epIdx > -1) {
+          episode = {...state[epIdx], id, seriesId, guid, page, loading: false, loaded: true};
+          episode[metricsPropertyName] = metrics;
+          newState = [...state.slice(0, epIdx), episode, ...state.slice(epIdx + 1)];
```

Comment:
Is there a reason to store the episodes in an array instead of an object (keyed by guid)?

Reply:
Fewer and smaller code changes. I wouldn't have the go ahead to change any of this stuff if the loading issues didn't make the release.